Dev/borealis io updates

mardet987 opened the pull request:
This update is for a new version of Borealis with some added file fields. I wanted to maintain that pydarn could read older Borealis files (v0.4 and lower) though, so I made the reader search for the version and then determine the format according to that version. 

I also updated restructuring by moving the restructuring functions into the format class. It makes sense here because these methods (site_to_array and array_to_site) use the fields from the format class anyways. So, restructure_borealis file is gone but borealis_formats file has really blown up. The formats are children of the previous format, and all are based on a base format class where I have placed some of the key functions so that there is less repeated code. 

This also removed some repeated code in the borealis_array and borealis_site read and write classes. There is no longer read_bfiq, read_rawacf, etc. there is just read_file, which just reads according to the format_class of the reader, which is determined in the init. 

I hope this is clear; although borealis_formats is somewhat confusing at first, it will make any future updates more straightforward as we can just update the fields necessary by creating a new child class. 

This has been rebased to master (release 1.0.0), linted to PEP8 standard and tested on both Borealisv0.4 and Borealis v0.5 files. There are some commits listed in the commit history that should've also been removed in the rebase - instead to fix it I have just erased those few changes with another couple of commits so that in the comparison you should only see my changes. (the rebase was extensive - removing hundreds of commits, so I would prefer not to redo it). 

Let me know if there are any questions!

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -98,10 +97,30 @@ def __init__(self, filename: str, borealis_filetype: str):
+        else:
+            self._borealis_version = version
+
+        self._format_class = borealis_formats.borealis_versions[
```
```suggestion
        self._format = borealis_formats.versions[
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -98,10 +97,30 @@ def __init__(self, filename: str, borealis_filetype: str):
+            self._borealis_version = version
+
+        self._format_class = borealis_formats.borealis_versions[
+                self.borealis_version][self.borealis_filetype]
```
```suggestion
                self.version][self.filetype]
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -137,9 +156,25 @@ def records(self):
-        return BorealisRestructureUtilities.borealis_array_to_site_dict(
-                                            self.filename, self.arrays,
-                                            self.borealis_filetype)
+        if self.format_class.is_restructureable():
```
```suggestion
        if self.format.is_restructureable()
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -137,9 +156,25 @@ def records(self):
-                                            self.borealis_filetype)
+        if self.format_class.is_restructureable():
+            try:
+                records = self.format_class._array_to_site(self.arrays)
```
```suggestion
                records = self.format._array_to_site(self.arrays)
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -137,9 +156,25 @@ def records(self):
+                records = self.format_class._array_to_site(self.arrays)
+                BorealisUtilities.check_records(
+                    self.filename, records,
+                    self.format_class.site_single_element_types(),
```
```suggestion
                    self.format.site_single_element_types(),
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -137,9 +156,25 @@ def records(self):
+                raise borealis_exceptions.BorealisRestructureError(
+                    'Arrays from {}: Error restructuring {} from array to site'
+                    ' style: {}'.format(self.filename,
+                                        self.format_class.__name__, e)) from e
```
``` suggestion
                                        self.format.__name__, err)) from err
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -137,9 +156,25 @@ def records(self):
+                    self.filename, records,
+                    self.format_class.site_single_element_types(),
+                    self.format_class.site_array_dtypes())
+            except Exception as e:
```
```suggestion
            except Exception as err:
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -137,9 +156,25 @@ def records(self):
+            raise borealis_exceptions.BorealisRestructureError(
+                'Arrays from {}: File format {} not recognized as '
+                'restructureable from site to array style or vice versa.'
+                ''.format(self.filename, self.format_class.__name__))
```
``` suggestion
                ''.format(self.filename, self.format.__name__))
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
         return self._arrays
 
+    @property
+    def borealis_version(self):
```
``` suggestion
    def version(self):
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
+        The version of the file, taken from the 'borealis_git_hash' field, in
+        the init.
+        """
+        return self._borealis_version
```
``` suggestion
        return self.version
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -73,6 +71,7 @@ class BorealisArrayRead():
     record_names: list(str)
```
No description for `self.format` or `self.format_class`

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
+        """
+        The format class used for the file, from the borealis_formats module.
+        """
+        return self._format_class
```
``` suggestion
        return self._format
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
         -------
         arrays: dict
-            borealis data dictionary. Keys are data field names and
+            borealis data dictionary of arrays. Keys are data field names and
```
``` suggestion
            Borealis data dictionary of arrays. Keys are data field names and
```

Might have missed some of these but do you prefer borealis or Borealis?

mardet987 replied:
changed all to Borealis with capital.

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
-                                   dataset_types, unshared_fields)
-        return self._arrays
+        pydarn_log.info("Reading Borealis {} {} file: {}"
+                        "".format(self.borealis_version,
```
``` suggestion 
                        "".format(self.version,
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
-        return self._arrays
+        pydarn_log.info("Reading Borealis {} {} file: {}"
+                        "".format(self.borealis_version,
+                                  self.borealis_filetype, self.filename))
```
``` suggestion
                                  self.filetype, self.filename))
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
-    def read_rawacf(self) -> dict:
-        """
-        Reads Borealis rawacf file that has been structured into arrays.
+        attribute_types = self.format_class.array_single_element_types()
```
``` suggestion
        attribute_types = self.format.array_single_element_types()
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
-        """
-        Reads Borealis rawacf file that has been structured into arrays.
+        attribute_types = self.format_class.array_single_element_types()
+        dataset_types = self.format_class.array_array_dtypes()
```
``` suggestion
        dataset_types = self.format.array_array_dtypes()
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
-        Reads Borealis rawacf file that has been structured into arrays.
+        attribute_types = self.format_class.array_single_element_types()
+        dataset_types = self.format_class.array_array_dtypes()
+        unshared_fields = self.format_class.unshared_fields()
```
``` suggestion
        unshared_fields = self.format.unshared_fields()
```

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +184,43 @@ def arrays(self):
-        unshared_fields = \
-            borealis_formats.BorealisAntennasIq.unshared_fields + \
-            borealis_formats.BorealisAntennasIq.array_only_fields
         self._read_borealis_arrays(attribute_types, dataset_types,
```
``` suggestion
        self._read_arrays(attribute_types, dataset_types,
```

mardet987 replied:
I prefer just protected for most things here. I don't want to limit users that want to write subclasses, for example, to not be able to access this directly without the name mangling. Although I don't use this function in any subclass, there are others that I do use in subclasses.

mts299 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -344,6 +321,24 @@ def __init__(self, filename: str, borealis_arrays: dict,
         self.filename = filename
         self._arrays = borealis_arrays
```
``` suggestion
        self._arrays = arrays
```

mardet987 commented on pydarn/io/borealis/borealis.py:
```diff
@@ -189,6 +188,14 @@ def arrays(self):
+        The Borealis version that created this data. May impact the
+        fields included in the file.
+        """
+        return self._reader._borealis_version
```
This version refers to the version of the Borealis software package. I'll update the doc for that also, but I think it might not be as clear if it's just version, since it could be version of multiple things. It kinds of reads as if there are multiple versions of the reader.

mts299 replied:
maybe `software_version` then? Again the package and class name is Borealis already so people may already know they are working with Borealis code related things.

mardet987 replied:
agreed replace with software version, will do.

mardet987 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -73,6 +71,7 @@ class BorealisArrayRead():
     record_names: list(str)
     records: dict
     arrays: dict
+    borealis_version : str
```
My preference would be to leave the borealis in front for anything that is defined by the Borealis software package; for filetypes this refers to filetypes specific to Borealis, like 'antennas_iq', so things that people may never have heard of before. Filetype could even mean 'hdf5' to an unknowing reader, so I think having the borealis in front gives some context that this is Borealis-specific.

mts299 replied:
Okay, just it already has Borealis in front of it with the package and class name being BorealisX so this just seemed redundant now.

mts299 commented on pydarn/io/borealis/borealis_formats.py:
```diff
@@ -246,27 +353,428 @@ def array_single_element_types(cls):
+        return num_lags
+
+    @staticmethod
+    def reshape_site_arrays(records: OrderedDict) -> OrderedDict:
```
So I feel this abusing inheritance since you are overwriting a lot the methods and they are classmethods or staticmethods. 
Here is a link explaining when to use inheritance: https://www.thoughtworks.com/insights/blog/composition-vs-inheritance-how-choose
Also determining when to use instance, static, and class: https://www.makeuseof.com/tag/python-instance-static-class-methods/

mts299 commented on pydarn/io/borealis/borealis_formats.py:
```diff
@@ -246,27 +353,428 @@ def array_single_element_types(cls):
+
+    @classmethod
+    def single_element_types(cls):
+        return {
```
Could this not be an attribute of the class?

mts299 commented on pydarn/io/borealis/borealis_formats.py:
```diff
@@ -246,27 +353,428 @@ def array_single_element_types(cls):
+
+    @classmethod
+    def array_dtypes(cls):
+        return {
```
Again could this be an attribute of class attribute so you can call like a variable?

mts299 commented on pydarn/io/borealis/borealis_site.py:
```diff
@@ -399,97 +402,60 @@ def arrays(self):
+                    self.format.array_array_dtypes(),
+                    self.format.unshared_fields())
+            except Exception as e:
+                raise borealis_exceptions.BorealisRestructureError(
```
add to raises

mts299 commented on test/unit/test_borealis.py:
```diff
@@ -36,18 +36,40 @@
+borealis_array_antennas_iq_file = \
+    "/home/marci/data/borealis-v05/20200303.2143.00.sas.0.antennas_iq.hdf5"
+
+# # Site Test files
```
can we delete the commented out files?

mardet987 commented on pydarn/io/borealis/borealis_array.py:
```diff
@@ -149,101 +185,43 @@ def arrays(self):
-        """
-        Reads Borealis antennas_iq file that has been structured into arrays.
+        attribute_types = self.format.array_single_element_types()
+        dataset_types = self.format.array_array_dtypes()
```
added a See Also here for the BaseFormat, where these are defined and described

mardet987 commented:
Hi @mts299. I think I've addressed your concerns with lots of extra documentation. To do this, I separated out BaseFormat into its own file, which is now over 1000 lines itself with all the docs. I've made sure to refer to the Borealis readthedocs (which will be updated with the sister PR for these updates on Borealis is merged into master). I've also written a note on the design concept and tried to separate the methods into clear categories (under BaseFormat documentation) to show how they are used. 

I updated the unit tests so that they automatically test both v0.4 and v0.5 files, reran the tests, and everything passed.

mts299 commented:
Looks good! Seems like there are some conflicts with your convert file if you want to check that out. 

Can you get someone to test it out? Maybe @Cooper10 @Devin-Huyghebaert 
Then I think it is good to merge :)

mardet987 commented:
Thanks @mts299 ! I fixed the merge conflicts.

Cooper10 commented:
I tested BorealisRead() on a .rawacf.hdf5 file and works fine:

mts299 commented:
@mardet987  is the test you were looking for? Wasn't sure if @Cooper10 tested on the new versioned type files?

Cooper10 commented:
I just tested a v0.5 file and it works.
'''
filename = "20200303.2143.00.sas.0.bfiq.hdf5"
borealis_reader = pydarn.BorealisRead(filename, 'bfiq')
print(borealis_reader.borealis_file_structure)      # returns array
'''
